core/src/main/java/org/ehcache/events/CacheEventNotificationServiceImpl.java

```diff
@@ -99,21 +139,51 @@ public void registerCacheEventListener(CacheEventListener<? super K, ? super V>
    */
   @Override
   public void deregisterCacheEventListener(CacheEventListener<? super K, ? super V> listener) {
-    if (!registeredListeners.remove(new EventListenerWrapper(listener,
-        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class)))) {
-      throw new IllegalStateException("Unknown cache event listener: " + listener);
+    EventListenerWrapper wrapper = new EventListenerWrapper(listener,
+        EventFiring.ASYNCHRONOUS, EventOrdering.UNORDERED, EnumSet.allOf(EventType.class));
+    boolean removed = false;
+    for (EventListenerWrapper listenerWrapper : aSyncListenersSet) {
+      if(listenerWrapper.equals(wrapper)) {
+        aSyncListenersSet.remove(listenerWrapper);
+        removed = true;
+        if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+          orderedListenerCount.decrementAndGet();
+        }
+        break;
+      }
     }
-    if (!hasListeners()) {
+    if (!removed) {
+      for (EventListenerWrapper listenerWrapper : syncListenersSet) {
+        if(listenerWrapper.equals(wrapper)) {
+          syncListenersSet.remove(listenerWrapper);
+          removed = true;
+          if(listenerWrapper.config.orderingMode() == EventOrdering.ORDERED) {
+            orderedListenerCount.decrementAndGet();
+          }
+          break;
+        }
+      }
+    }
+    synchronized (this) {
+      if (!hasListeners()) {
+        if (eventThreadLocal instanceof EventThreadLocalImpl) {
+          eventThreadLocal = new NoOpEventThreadLocalImpl();
+        }
+      }
       store.disableStoreEventNotifications();
```

Comment:
This is now outside of the `hasListeners()` check which means the first deregister will cause store notifications to be disable. Not what we want.